### ISSUE TITLE
Possible to retrieve peer info?

### ISSUE THREAD

**Issue by eberkund** [bufbuild/connect-go#357]
With `grpc-go` I can get peer info like so:

```go
p, ok := peer.FromContext(ctx)
if !ok {
    return nil, errors.New("could not get peer from context")
}
addr := p.Addr.String()
```

Is it possible to do this with `connect-go`?

I am trying to get the IP that the request is originating from.

Thanks

**Comment by akshayjshah** [comment]
Currently, it's possible but not super-simple. You'd need to write a small piece of `net/http` middleware to attach the HTTP [`Request.RemoteAddr`](https://pkg.go.dev/net/http#:~:text=Trailer%20Header%0A%0A%09//-,RemoteAddr,-allows%20HTTP%20servers) to the context, then pull it off the context and include it in your logs.

I'm open to exposing this in handlers, but I'd like to see what else people expect to do with this information. For example, we don't have any current plans to expose an equivalent to the `grpc.Peer` call options.

**Comment by clly** [comment]
Some potential usages I can think of include per IP rate limiting, caching, or naive authorization.

It might be nicer to include it in the request object like the headers are. It makes them more explicit, less magical, and easier to use.

**Comment by eberkund** [comment]
A middleware like @akshayjshah suggest indeed works for anyone who comes across this issue. This is how I implemented it:

```go
type remoteAddrKey struct{}

func addRemoteAddr(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		ctx := context.WithValue(r.Context(), remoteAddrKey{}, r.RemoteAddr)
		req := r.WithContext(ctx)
		next.ServeHTTP(w, req)
	})
}
```

Make sure you wrap your handler in the middleware (like any other HTTP middleware in Go):

```go
mux.Handle(path, addRemoteAddr(handler))
```

And then in your handler:

```go
remoteAddr, ok := ctx.Value(remoteAddrKey{}).(string)
```

**Comment by eberkund** [comment]
But besides the option above, is there a way to get the remote address without the middleware? I saw the IP address is already in the context when I was playing around with my debugger but I just didn't get to a point of figuring out how to extract it.

**Comment by akshayjshah** [comment]
> I saw the IP address is already in the context when I was playing around with my debugger but I just didn't get to a point of figuring out how to extract it.

Nothing in `connect-go` is putting the IP onto the context, and AFAIK nothing in `net/http` is either.

**Comment by akshayjshah** [comment]
Opened a PR that exposes this info on `Request`, `AnyRequest`, `StreamingClientConn`, `StreamingHandlerConn`, and the relevant wrappers. @clly, we're definitely not attaching anything to the context!

Usage looks like this:

```go
type PingServer struct {}

func (p *PingServer) Ping(
  ctx context.Context,
  request *connect.Request[pingv1.PingRequest],
) (*connect.Response[pingv1.PingResponse], error) {
  fmt.Println(request.Peer().Addr) // <-- this is new!
  return connect.NewResponse(&pingv1.PingResponse{
    Number: request.Msg.Number,
  }), nil
}
```

**Comment by akshayjshah** [comment]
Released in v0.5.0!
